src/crypto/mod.rs

```diff
@@ -0,0 +1,293 @@
+    /// `i + 1` is key share number `i`.
+    coeff: Vec<E::G1>,
+}
+
+impl<E: Engine> PublicKeySet<E> {
+    /// Returns the threshold `t`: any set of `t + 1` signature shares can be combined into a full
+    /// signature.
+    pub fn threshold(&self) -> usize {
+        self.coeff.len() - 1
+    }
+
+    /// Returns the public key.
+    pub fn public_key(&self) -> PublicKey<E> {
+        PublicKey(self.coeff[0])
+    }
+
+    /// Returns the `i`-th public key share.
+    pub fn public_key_share<T>(&self, i: T) -> PublicKey<E>
+    where
+        T: Into<<E::Fr as PrimeField>::Repr>,
+    {
+        let mut x = E::Fr::one();
+        x.add_assign(&E::Fr::from_repr(i.into()).expect("invalid index"));
+        let mut pk = *self.coeff.last().expect("at least one coefficient");
+        for c in self.coeff.iter().rev().skip(1) {
+            pk.mul_assign(x);
+            pk.add_assign(c);
+        }
+        PublicKey(pk)
+    }
+
+    /// Verifies that the given signatures are correct and combines them into a signature that can
+    /// be verified with the main public key.
+    pub fn combine_signatures<'a, ITR, IND>(&self, items: ITR) -> Result<Signature<E>>
+    where
+        ITR: IntoIterator<Item = (&'a IND, &'a Signature<E>)>,
+        IND: Into<<E::Fr as PrimeField>::Repr> + Clone + 'a,
+    {
+        let sigs: Vec<_> = items
+            .into_iter()
+            .map(|(i, sig)| {
+                let mut x = E::Fr::one();
+                x.add_assign(&E::Fr::from_repr(i.clone().into()).expect("invalid index"));
+                (x, sig)
+            })
+            .collect();
+        if sigs.len() < self.coeff.len() {
+            return Err(ErrorKind::NotEnoughShares.into());
+        }
+        let mut result = E::G2::zero();
+        let mut indexes = Vec::new();
+        for (x, sig) in sigs.iter().take(self.coeff.len()) {
+            if indexes.contains(x) {
+                return Err(ErrorKind::DuplicateEntry.into());
+            }
+            indexes.push(x.clone());
+            // Compute the value at 0 of the Lagrange polynomial that is `0` at the other data
+            // points but `1` at `x`.
+            let mut l0 = E::Fr::one();
+            for (x0, _) in sigs.iter().take(self.coeff.len()).filter(|(x0, _)| x0 != x) {
+                let mut denom = *x0;
+                denom.sub_assign(x);
+                l0.mul_assign(x0);
+                l0.mul_assign(&denom.inverse().expect("indices are different"));
+            }
+            let mut summand = sig.0;
+            summand.mul_assign(l0);
+            result.add_assign(&summand);
+        }
+        Ok(Signature(result))
+    }
+}
+
+/// A secret key and an associated set of secret key shares.
+pub struct SecretKeySet<E: Engine> {
+    /// The coefficients of a polynomial whose value at `0` is the "master key", and value at
+    /// `i + 1` is key share number `i`.
```

Comment:
I see that you refer to coefficients by their positions in the `Vec`. Wouldn't it be more descriptive to refer here to the polynomial instead of the vector that stores the coefficients? E.g., "the master key is the 0-th coefficient of the polynomial and, for 0 < i <= N, the i-th coefficient of the polynomial is the (i-1)-th key share".

Reply:
But the `(i - 1)`-th key share isn't the `i`-th coefficient; it's the value at `i`:

```rust
coeff[0] + i * coeff[1] + i * i * coeff[2] + …
```

Not sure what would be a better formulation.